File: visible_tree_node.js
class Node {
    constructor(val, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

function visibleTreeNode(root) {
    // WRITE YOUR BRILLIANT CODE HERE
    const dfs = function(node, maxVal) {
        if (!node) return 0;
        maxVal = Math.max(maxVal, node.val)
        return (node.val >= maxVal) + dfs(node.left, maxVal) + dfs(node.right, maxVal);
    }
    return dfs(root, root.val);
}


/**
 * Algomonster solution
 */
 function dfs(root, max_sofar) {
    if (!root) return 0;
    let total = 0;
    if (root.val >= max_sofar) total++;

    // max_sofar for child node is the larger of previous max and current node val
    total += dfs(root.left, Math.max(max_sofar, root.val));
    total += dfs(root.right, Math.max(max_sofar, root.val));

    return total;
}

function visibleTreeNode(root) {
    // start max_sofar with smallest number possible so any value root has is greater than it
    return dfs(root, Number.NEGATIVE_INFINITY);
}

File: python/visible_tree_node.py
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def visible_tree_node(root: Node) -> int:
    # WRITE YOUR BRILLIANT CODE HERE
    def dfs(root, currMax):
        if not root:
            return 0
        if root.val >= currMax:  
            return dfs(root.left, root.val) + dfs(root.right, root.val) + 1
        return dfs(root.left, currMax) + dfs(root.right, currMax)
        
    
    return dfs(root, root.val)